kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/CachingKms.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption;
+
+import java.time.Duration;
+import java.util.concurrent.CompletionStage;
+
+import javax.crypto.SecretKey;
+
+import com.github.benmanes.caffeine.cache.AsyncLoadingCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+import io.kroxylicious.kms.service.DekPair;
+import io.kroxylicious.kms.service.Kms;
+import io.kroxylicious.kms.service.Serde;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Kms implementation that caches results from a delegate where appropriate.
+ * We can cache resolved aliases for some time, and we can cache decrypted EDEKs
+ * indefinitely.
+ * @param <K>
+ * @param <E>
+ */
+public class CachingKms<K, E> implements Kms<K, E> {
+    private final Kms<K, E> delegate;
+    private final AsyncLoadingCache<E, SecretKey> decryptDekCache;
+    private final AsyncLoadingCache<String, K> resolveAliasCache;
+
+    private CachingKms(Kms<K, E> delegate,
+                       long decryptDekCacheMaxSize,
+                       Duration decryptDekExpireAfterAccess,
+                       long resolveAliasCacheMaxSize,
+                       Duration resolveAliasExpireAfterWrite) {
+        this.delegate = delegate;
+        decryptDekCache = buildDecryptedDekCache(delegate, decryptDekCacheMaxSize, decryptDekExpireAfterAccess);
+        resolveAliasCache = buildResolveAliasCache(delegate, resolveAliasCacheMaxSize, resolveAliasExpireAfterWrite);
+    }
+
+    @NonNull
+    public static <A, B> Kms<A, B> caching(Kms<A, B> delegate,
+                                           long decryptDekCacheMaxSize,
+                                           Duration decryptDekExpireAfterAccess,
+                                           long resolveAliasCacheMaxSize,
+                                           Duration resolveAliasExpireAfterWrite) {
+        return new CachingKms<>(delegate, decryptDekCacheMaxSize, decryptDekExpireAfterAccess, resolveAliasCacheMaxSize, resolveAliasExpireAfterWrite);
+    }
+
+    @NonNull
+    private static <K, E> AsyncLoadingCache<String, K> buildResolveAliasCache(Kms<K, E> delegate, long maxSize, Duration expireAliasResolutionAfterWrite) {
+        return Caffeine.newBuilder().maximumSize(maxSize).expireAfterWrite(expireAliasResolutionAfterWrite)
```

Comment:
we should also probably put a `refreshAfterWrite` less than the expiry on it as well so that an actively used alias will be refreshed in the background.

https://github.com/ben-manes/caffeine/wiki/Refresh

> In contrast to expireAfterWrite, refreshAfterWrite will make a key eligible for refresh after the specified duration, but a refresh will only be actually initiated when the entry is queried. So, for example, you can specify both refreshAfterWrite and expireAfterWrite on the same cache so that the expiration timer on an entry isn't blindly reset whenever an entry becomes eligible for a refresh. If an entry isn't queried after it comes eligible for refreshing, it is allowed to expire.

Reply:
done, have hardcoded it to a 10 minute after write eviction with an 8 minute after write refresh